src/cargo/ops/cargo_update.rs

```diff
@@ -207,6 +214,251 @@ pub fn print_lockfile_changes(
+        trace!(
+            "skipping dependency `{}` because the version requirement didn't change",
+            name
+        );
+        return Ok(dependency);
+    };
+
+    let upgrade_message = format!("{} {} -> {}", name, current, new_req_string);
+    trace!(upgrade_message);
+
+    if upgrade_messages.insert(upgrade_message.clone()) {
+        opts.gctx
+            .shell()
+            .status_with_color("Upgrading", &upgrade_message, &style::GOOD)?;
+    }
+
+    upgrades.insert(name.to_string(), latest.clone());
+
+    let req = OptVersionReq::Req(VersionReq::parse(&latest.to_string())?);
+    let mut dep = dependency.clone();
+    dep.set_version_req(req);
+    Ok(dep)
+}
+
+/// Update manifests with upgraded versions, and write to disk. Based on cargo-edit.
+/// Returns true if any file has changed.
+pub fn write_manifests(
+    ws: &Workspace<'_>,
+    opts: &UpdateOptions<'_>,
+    upgrades: &HashMap<String, Version>,
+) -> CargoResult<bool> {
+    if upgrades.is_empty() {
+        return Ok(false);
+    }
+
+    let mut any_file_has_changed = false;
+
+    let manifest_paths = std::iter::once(ws.root_manifest())
+        .chain(ws.members().map(|member| member.manifest_path()))
+        .collect::<Vec<_>>();
+
+    for manifest_path in manifest_paths {
+        trace!(
+            "updating TOML manifest at `{:?}` with upgraded dependencies",
+            manifest_path
+        );
+
+        let crate_root = manifest_path
+            .parent()
+            .expect("manifest path is absolute")
+            .to_owned();
+
+        let mut local_manifest = LocalManifest::try_new(&manifest_path)?;
+        let mut manifest_has_changed = false;
+
+        for dep_table in local_manifest.get_dependency_tables_mut() {
+            for (mut dep_key, dep_item) in dep_table.iter_mut() {
+                let dep_key_str = dep_key.get();
+                let dependency = match crate::util::toml_mut::dependency::Dependency::from_toml(
+                    &manifest_path,
+                    dep_key_str,
+                    dep_item,
+                ) {
+                    Ok(dependency) => dependency,
+                    Err(err) => {
+                        opts.gctx
+                            .shell()
+                            .warn(&format!("ignoring {dep_key}, unsupported entry: {err}"))?;
+                        continue;
+                    }
+                };
+
+                let Some(current) = dependency.version() else {
+                    trace!("skipping dependency without a version: {}", dependency.name);
+                    continue;
+                };
+
+                let Some(latest) = upgrades.get(&dependency.name) else {
+                    trace!(
+                        "skipping dependency without an upgrade: {}",
+                        dependency.name
+                    );
+                    continue;
+                };
```

Comment:
What about the following cases
- A `path` dependency with a `version` specified
- A `path` dependency with a `version` specified **and** a version-only dependency with the same name

Reply:
The first case is already covered:

```toml
bar  =  { path  =  "../bar", version  =  "0.0.1" }  # Comment
```

Reply:
I added a test case, `conflicting-source-type`. Hope that is what you had in mind.

Reply:
I worry that use of `package` for renaming could make it we aren't testing the case I'm interested in.

Also, its not quite conflicting.

Reply:
I changed the names to something more accurate than `conflicting-*`. And I updated the multiple source types test to not do renaming.